firefox_ui_tests/remote/security/test_safe_browsing_warning_pages.py

```diff
@@ -68,31 +68,28 @@ def check_get_me_out_of_here_button(self, unsafe_page):
         self.wait_for_condition(lambda mn: self.browser.default_homepage in mn.get_url())
 
     def check_report_button(self, unsafe_page):
+        # Get the URL of the support site for phishing and malware. This may result in a redirect.
+        with self.marionette.using_context('chrome'):
+            url = self.marionette.execute_script("""
+              Components.utils.import("resource://gre/modules/Services.jsm");
+              return Services.urlFormatter.formatURLPref("app.support.baseURL")
+                                                         + "phishing-malware";
+            """)
+
         button = self.marionette.find_element(By.ID, "reportButton")
         button.click()
 
         # Wait for the button to become stale, then wait for page load
-        # so we can verify the url
+        # so we can verify the url even if a redirect happens
         self.wait_for_condition(expected.element_stale(button))
 
-        # Get the base URL to check; this will result in a redirect.
-        with self.marionette.using_context('chrome'):
-            if 'its-a-trap' in unsafe_page:
-                url = self.marionette.execute_script("""
-                  Components.utils.import("resource://gre/modules/Services.jsm");
-                  return Services.urlFormatter.formatURLPref("app.support.baseURL")
-                                                             + "phishing-malware";
-                """)
-            else:
-                url = self.marionette.execute_script("""
-                  Components.utils.import("resource://gre/modules/Services.jsm");
-                  return Services.urlFormatter.formatURLPref(
-                  "browser.safebrowsing.malware.reportURL") + arguments[0];
-                """, script_args=[unsafe_page])
+        # TODO: Bug 1140470: use replacement for mozmill's waitforPageLoad
+        self.wait_for_condition(lambda mn: mn.execute_script("""
+          return document.readyState == 'complete';
+        """))
```

Comment:
The problem is the part where you called `get_final_url()`. This helper method loads the specified url and waits for the loading be completed. But exactly that unloads the webpage which was loaded by the button click. So your `Wait().until()` call operates on the exact same loaded URL and always passes in the first iteration. With the assert it first retrieves the current URL, and then constructs the final URL. That way we will notice if the wanted URL has been loaded or not.